jsx-sort-props rule is not working as it should

Rule configuration:
```js
'react/jsx-sort-props': [
  'error',
  {
    noSortAlphabetically: false,
    ignoreCase: true,
    callbacksLast: false,
    shorthandLast: true,
    reservedFirst: true,
  },
],
```

As a result `callbacksLast`, `shorthandLast` options are overridden by `noSortAlphabetically`, so props are sorting only by alphabet ignoring shorthands at the end. Also, how can I reach this sort order:

- object spread props
- reserved props
- common props (html attributes, etc.)
- callback props
- shorthand/boolean props

Example:
```jsx
<button
  {...rest}
  key={1}
  type="button"
  onClick={() => {}}
  disabled
/>
```

Hey I just stumbled over this as well. I figured I just add some requested data:

## Example Code + Config

### Code
```javascript
import React from 'react';

export const Test = () => (
  <p z="test" a />
);
```

### Config

```json5
{
  "root": true,
  "extends": ["eslint:recommended", "plugin:react/recommended"],
  "parser": "babel-eslint",
  "parserOptions": {
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true,
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": [
    "react"
  ],
  "rules": {
    "react/jsx-sort-props": [2, {
      "reservedFirst": true,
      "shorthandLast": true
    }
  }
}
```

## Expected Output

Note: I run `node_modules/.bin/eslint <file>` to produce the following output.
```eslint
```

## Actual Output

```eslint
/Users/kfle/Documents/Programming/work/OetkerDigital/backen/test.js
  4:15  error  Props should be sorted alphabetically  react/jsx-sort-props

✖ 1 problem (1 error, 0 warnings)
  1 error, 0 warnings potentially fixable with the `--fix` option.
```

## "Interesting" Observation

If I only supply the `shorthandLast` linting rule I get the expected result (also the expected output, if I incorrectly sort the shorthand prop before any others).

Example working config for only shorthand rule:
```json5
  ...
  "rules": {
    "react/jsx-sort-props": [2, {
      "shorthandLast": true
    }]
  }
```

--- a/lib/rules/jsx-sort-props.js
+++ b/lib/rules/jsx-sort-props.js
@@ -245,15 +245,8 @@ module.exports = {
             const previousIsReserved = isReservedPropName(previousPropName, reservedList);
             const currentIsReserved = isReservedPropName(currentPropName, reservedList);
 
-            if ((previousIsReserved && currentIsReserved) || (!previousIsReserved && !currentIsReserved)) {
-              if (!noSortAlphabetically && currentPropName < previousPropName) {
-                context.report({
-                  node: decl,
-                  message: 'Props should be sorted alphabetically',
-                  fix: generateFixerFunction(node, context, reservedList)
-                });
-                return memo;
-              }
+            if (previousIsReserved && !currentIsReserved) {
+              return decl;
             }
             if (!previousIsReserved && currentIsReserved) {
               context.report({
@@ -261,8 +254,8 @@ module.exports = {
                 message: 'Reserved props must be listed before all other props',
                 fix: generateFixerFunction(node, context, reservedList)
               });
+              return memo;
             }
-            return decl;
           }
 
           if (callbacksLast) {
@@ -310,7 +303,7 @@ module.exports = {
             context.report({
               node: decl,
               message: 'Props should be sorted alphabetically',
-              fix: generateFixerFunction(node, context)
+              fix: generateFixerFunction(node, context, reservedList)
             });
             return memo;
           }

--- a/tests/lib/rules/jsx-sort-props.js
+++ b/tests/lib/rules/jsx-sort-props.js
@@ -88,6 +88,10 @@ const reservedFirstWithNoSortAlphabeticallyArgs = [{
   noSortAlphabetically: true,
   reservedFirst: true
 }];
+const reservedFirstWithShorthandLast = [{
+  reservedFirst: true,
+  shorthandLast: true
+}];
 const reservedFirstAsEmptyArrayArgs = [{
   reservedFirst: []
 }];
@@ -149,6 +153,10 @@ ruleTester.run('jsx-sort-props', rule, {
     {
       code: '<div ref="r" dangerouslySetInnerHTML={{__html: "EPR"}} key={0} children={<App />} b a c />',
       options: reservedFirstWithNoSortAlphabeticallyArgs
+    },
+    {
+      code: '<App key="key" c="c" b />',
+      options: reservedFirstWithShorthandLast
     }
   ],
   invalid: [
@@ -231,6 +239,16 @@ ruleTester.run('jsx-sort-props', rule, {
       errors: 3
     },
     {
+      code: '<App key="key" b c="c" />',
+      errors: [expectedShorthandLastError],
+      options: reservedFirstWithShorthandLast
+    },
+    {
+      code: '<App ref="ref" key="key" isShorthand veryLastAttribute="yes" />',
+      errors: [expectedError, expectedShorthandLastError],
+      options: reservedFirstWithShorthandLast
+    },
+    {
       code: '<App a z onFoo onBar />;',
       errors: [expectedError],
       options: callbacksLastArgs
@@ -293,7 +311,7 @@ ruleTester.run('jsx-sort-props', rule, {
       code: '<App dangerouslySetInnerHTML={{__html: "EPR"}} e key={2} b />',
       options: reservedFirstAsBooleanArgs,
       output: '<App key={2} b dangerouslySetInnerHTML={{__html: "EPR"}} e />',
-      errors: [expectedReservedFirstError]
+      errors: [expectedReservedFirstError, expectedError]
     },
     {
       code: '<App key={3} children={<App />} />',

